src/libraries/Microsoft.PowerFx.Interpreter/PatchFunction.cs

```diff
@@ -0,0 +1,182 @@
+         * => Trying to patch on an Immutable object. We should get an error.
+         */
+
+    internal abstract class PatchAndValidateRecordFunctionBase : BuiltinFunction
+    {
+        public PatchAndValidateRecordFunctionBase(DPath theNamespace, string name, TexlStrings.StringGetter description, FunctionCategories fc, DType returnType, BigInteger maskLambdas, int arityMin, int arityMax, params DType[] paramTypes)
+           : base(theNamespace, name, /*localeSpecificName*/string.Empty, description, fc, returnType, maskLambdas, arityMin, arityMax, paramTypes)
+        {
+        }
+
+        public PatchAndValidateRecordFunctionBase(string name, TexlStrings.StringGetter description, FunctionCategories fc, DType returnType, BigInteger maskLambdas, int arityMin, int arityMax, params DType[] paramTypes)
+            : this(DPath.Root, name, description, fc, returnType, maskLambdas, arityMin, arityMax, paramTypes)
+        {
+        }
+
+        public override bool CheckInvocation(TexlBinding binding, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(binding);
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+
+            var isValid = CheckInvocation(args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+
+            return isValid;
+        }
+    }
+
+    // Patch( Record1, Record2 [, …] )
+    internal class PatchRecordFunction : PatchAndValidateRecordFunctionBase, IAsyncTexlFunction
+    {
+        public override bool IsSelfContained => false;
+
+        public PatchRecordFunction()
+            : base("Patch", TexlStrings.AboutPatch, FunctionCategories.Table | FunctionCategories.Behavior, DType.EmptyRecord, 0, 2, int.MaxValue, DType.EmptyRecord, DType.EmptyRecord)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg };
+            yield return new[] { TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg, TexlStrings.PatchChangeRecordsArg };
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            if (arity > 2)
+            {
+                return GetGenericSignatures(arity, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg);
+            }
+
+            return base.GetSignatures(arity);
+        }
+
+        public Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
```

Comment:
Why is the implementation defined in the same file as the front-end/type checking? Why are there no changes to Library.cs? Why does these functions not have standard error handling wrappers?